qdata/parser.py

```diff
@@ -311,57 +425,106 @@ def mixedlist(self, *args):
         return ParsedList(Lists.MIXEDLIST, wires)
 
     def argument(self, *args):
-        # <id> or
-        # <id>[<nninteger>]
+        """An argument is either an id, or an id with a non-negative integer index."""
         return unpack(args)
 
     def explist(self, *args):
+        """Either a single expression, or a comma separated list of expressions"""
         flat_list = flatten(*args)
         return ParsedList(Lists.EXPLIST, flat_list)
 
     def exp(self, *args):
+        """A mathematica expression.
+
+        This includes:
+
+        * A real value
+        * A non-negative integer
+        * Mathematical constants (currently just pi)
+        * A symbolic parameter (simply id, so must always start with a lowercase character)
+        * Parenthesis
+        * Unary arithmetic operators (-, unaryop)
+        * Binary arithmetic operators (+, -, *, /, ^)
+
+        .. code-block:: text
+
+            real
+            nninteger
+            PI
+            id -> parameter
+            exp "+" exp -> add
+            exp "-" exp -> subtract
+            exp "*" exp -> multiply
+            exp "/" exp -> divide
+            "-" exp -> negate
+            exp "^" exp -> binary_exp
+            "(" exp ")"
+            unaryop "(" exp ")"
+        """
+        # TODO: Check what happens if a negative integer is specified. This will likely
+        # result in a float being returned, which is not ideal (larger memory requirements,
+        # plus the potential for propagation of precision error).
+
+        # TODO: support parameters that start with a capital letter?
```

Comment:
:pray:

Reply:
the file `qelibc.inc` actually has a declaration for a gate `cx` in terms of `CX` :laughing: